File: backend/app/models/opencode.py
"""
OpenCode Agent 审计集成模型

支持:
- OpenCode 审计项目管理（进程级隔离）
- Skills 库（可注入的审计技能）
- MCP 工具配置
- 问题评论与闭环流程
"""

import uuid
from datetime import datetime
from typing import Optional
from sqlalchemy import (
    Column, String, Integer, Boolean, Text, DateTime,
    ForeignKey, JSON, Date
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.db.base import Base


class OpenCodeServerStatus:
    """OpenCode 服务器状态"""
    STOPPED = "stopped"
    STARTING = "starting"
    RUNNING = "running"
    AUDITING = "auditing"
    STOPPING = "stopping"
    ERROR = "error"


class OpenCodeProject(Base):
    """
    OpenCode 审计项目

    每个项目对应一个独立的 OpenCode 服务进程，
    通过端口隔离实现多项目并行审计。
    """
    __tablename__ = "opencode_projects"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(
        String(36),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )

    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)

    # OpenCode 服务器配置
    code_path = Column(String(500), nullable=False)
    port = Column(Integer, nullable=True)
    status = Column(String(30), default=OpenCodeServerStatus.STOPPED, index=True)
    server_pid = Column(Integer, nullable=True)
    last_health_check = Column(DateTime(timezone=True), nullable=True)
    health_check_error = Column(Text, nullable=True)

    # 审计配置
    selected_skills = Column(JSON, nullable=True, default=list)
    selected_mcp_tools = Column(JSON, nullable=True, default=list)
    audit_config = Column(JSON, nullable=True)

    # 当前/最近审计任务
    current_agent_task_id = Column(
        String(36),
        ForeignKey("agent_tasks.id", ondelete="SET NULL"),
        nullable=True,
    )

    # 元数据
    created_by = Column(String(36), ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    # 关联关系
    project = relationship("Project", foreign_keys=[project_id])
    creator = relationship("User", foreign_keys=[created_by])
    audit_sessions = relationship(
        "OpenCodeAuditSession",
        back_populates="opencode_project",
        cascade="all, delete-orphan",
        order_by="OpenCodeAuditSession.created_at.desc()",
    )

    def __repr__(self):
        return f"<OpenCodeProject {self.name} [{self.status}]>"

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "project_id": self.project_id,
            "name": self.name,
            "description": self.description,
            "code_path": self.code_path,
            "port": self.port,
            "status": self.status,
            "server_pid": self.server_pid,
            "selected_skills": self.selected_skills or [],
            "selected_mcp_tools": self.selected_mcp_tools or [],
            "audit_config": self.audit_config,
            "current_agent_task_id": self.current_agent_task_id,
            "created_by": self.created_by,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }


class OpenCodeAuditSession(Base):
    """OpenCode 审计会话记录"""
    __tablename__ = "opencode_audit_sessions"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    opencode_project_id = Column(
        String(36),
        ForeignKey("opencode_projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    agent_task_id = Column(
        String(36),
        ForeignKey("agent_tasks.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )

    # 会话配置快照（审计时使用的 skills 和 MCP）
    skills_snapshot = Column(JSON, nullable=True)
    mcp_snapshot = Column(JSON, nullable=True)
    audit_config_snapshot = Column(JSON, nullable=True)

    # 会话状态
    status = Column(String(30), default="pending")
    opencode_session_id = Column(String(255), nullable=True)

    # 统计
    messages_count = Column(Integer, default=0)
    findings_count = Column(Integer, default=0)

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    completed_at = Column(DateTime(timezone=True), nullable=True)

    # 关联关系
    opencode_project = relationship("OpenCodeProject", back_populates="audit_sessions")

    def __repr__(self):
        return f"<OpenCodeAuditSession {self.id} [{self.status}]>"


class SkillCategory:
    """Skill 分类"""
    OWASP = "owasp"
    FRAMEWORK = "framework"
    BUSINESS = "business"
    COMPLIANCE = "compliance"
    CUSTOM = "custom"


class OpenCodeSkill(Base):
    """
    OpenCode 审计技能

    Skill 是可注入到 OpenCode 的结构化审计能力描述，
    通过系统提示词告诉 AI 关注哪类安全问题。
    """
    __tablename__ = "opencode_skills"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(100), nullable=False, unique=True)
    display_name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    category = Column(String(50), default=SkillCategory.CUSTOM, index=True)

    # Skill 核心内容
    system_prompt = Column(Text, nullable=False)
    tool_hints = Column(JSON, nullable=True)
    vulnerability_types = Column(JSON, nullable=True)
    severity_focus = Column(JSON, nullable=True)

    # 显示信息
    icon = Column(String(50), nullable=True)
    tags = Column(JSON, nullable=True, default=list)

    # 状态
    is_system = Column(Boolean, default=False)
    is_active = Column(Boolean, default=True)

    created_by = Column(String(36), ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    def __repr__(self):
        return f"<OpenCodeSkill {self.name}>"

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "display_name": self.display_name,
            "description": self.description,
            "category": self.category,
            "system_prompt": self.system_prompt,
            "tool_hints": self.tool_hints,
            "vulnerability_types": self.vulnerability_types or [],
            "severity_focus": self.severity_focus,
            "icon": self.icon,
            "tags": self.tags or [],
            "is_system": self.is_system,
            "is_active": self.is_active,
            "created_by": self.created_by,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }


class MCPTransportType:
    """MCP 传输类型"""
    HTTP = "http"
    STDIO = "stdio"
    SSE = "sse"


class MCPToolConfig(Base):
    """
    MCP 工具配置

    支持通过 Model Context Protocol 扩展 OpenCode 的工具能力。
    """
    __tablename__ = "mcp_tool_configs"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(100), nullable=False, unique=True)
    display_name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)

    # 连接配置
    transport_type = Column(String(20), default=MCPTransportType.HTTP)
    server_url = Column(String(500), nullable=True)
    command = Column(String(500), nullable=True)
    args = Column(JSON, nullable=True, default=list)
    env_vars = Column(JSON, nullable=True, default=dict)
    timeout_seconds = Column(Integer, default=30)

    # 工具能力描述
    capabilities = Column(JSON, nullable=True, default=list)
    icon = Column(String(50), nullable=True)

    # 状态
    is_active = Column(Boolean, default=True)
    is_system = Column(Boolean, default=False)
    last_test_status = Column(String(20), nullable=True)
    last_tested_at = Column(DateTime(timezone=True), nullable=True)

    created_by = Column(String(36), ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    def __repr__(self):
        return f"<MCPToolConfig {self.name} [{self.transport_type}]>"

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "display_name": self.display_name,
            "description": self.description,
            "transport_type": self.transport_type,
            "server_url": self.server_url,
            "command": self.command,
            "args": self.args or [],
            "env_vars": self.env_vars or {},
            "timeout_seconds": self.timeout_seconds,
            "capabilities": self.capabilities or [],
            "icon": self.icon,
            "is_active": self.is_active,
            "is_system": self.is_system,
            "last_test_status": self.last_test_status,
            "last_tested_at": self.last_tested_at.isoformat() if self.last_tested_at else None,
            "created_by": self.created_by,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }


class IssueCommentType:
    """问题评论类型（同时驱动状态机）"""
    NOTE = "note"
    CONFIRM = "confirm"
    REJECT = "reject"
    ASSIGN = "assign"
    FIX = "fix"
    VERIFY = "verify"
    REOPEN = "reopen"
    ACCEPT_RISK = "accept_risk"


class IssueComment(Base):
    """
    问题评论

    支持对 AgentFinding 的评论、确认、分配、关闭等操作，
    构建完整的问题闭环管理流程。
    """
    __tablename__ = "issue_comments"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    finding_id = Column(
        String(36),
        ForeignKey("agent_findings.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # 评论内容
    comment_type = Column(String(30), default=IssueCommentType.NOTE)
    content = Column(Text, nullable=False)

    # 状态变更记录
    from_status = Column(String(30), nullable=True)
    to_status = Column(String(30), nullable=True)

    # 作者信息
    author_id = Column(String(36), ForeignKey("users.id"), nullable=False)

    created_at = Column(DateTime(timezone=True), server_default=func.now())

    # 关联关系
    author = relationship("User", foreign_keys=[author_id])
    finding = relationship("AgentFinding", foreign_keys=[finding_id])

    def __repr__(self):
        return f"<IssueComment {self.comment_type} on {self.finding_id}>"

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "finding_id": self.finding_id,
            "comment_type": self.comment_type,
            "content": self.content,
            "from_status": self.from_status,
            "to_status": self.to_status,
            "author_id": self.author_id,
            "author_name": self.author.full_name if self.author else None,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }

File: frontend/src/shared/api/opencode.ts
/**
 * OpenCode Agent 审计 API 客户端
 */

import { apiClient } from "@/shared/api/serverClient";

// ─── Types ────────────────────────────────────────────────────────────────────

export type OpenCodeServerStatus =
  | "stopped"
  | "starting"
  | "running"
  | "auditing"
  | "stopping"
  | "error";

export interface OpenCodeProject {
  id: string;
  project_id?: string;
  name: string;
  description?: string;
  code_path: string;
  port?: number;
  status: OpenCodeServerStatus;
  server_pid?: number;
  selected_skills: string[];
  selected_mcp_tools: string[];
  audit_config?: Record<string, unknown>;
  current_agent_task_id?: string;
  created_by: string;
  created_at: string;
  updated_at?: string;
  skills_detail?: OpenCodeSkill[];
  mcp_tools_detail?: MCPToolConfig[];
}

export type SkillCategory = "owasp" | "framework" | "business" | "compliance" | "custom";

export interface OpenCodeSkill {
  id: string;
  name: string;
  display_name: string;
  description?: string;
  category: SkillCategory;
  system_prompt: string;
  tool_hints?: string[];
  vulnerability_types?: string[];
  severity_focus?: string[];
  icon?: string;
  tags?: string[];
  is_system: boolean;
  is_active: boolean;
  created_by?: string;
  created_at: string;
}

export type MCPTransportType = "http" | "stdio" | "sse";

export interface MCPToolConfig {
  id: string;
  name: string;
  display_name: string;
  description?: string;
  transport_type: MCPTransportType;
  server_url?: string;
  command?: string;
  args?: string[];
  env_vars?: Record<string, string>;
  timeout_seconds: number;
  capabilities?: string[];
  icon?: string;
  is_active: boolean;
  is_system: boolean;
  last_test_status?: "ok" | "failed";
  last_tested_at?: string;
  created_by?: string;
  created_at: string;
}

export interface OpenCodeAuditSession {
  id: string;
  agent_task_id?: string;
  status: string;
  findings_count: number;
  messages_count: number;
  created_at: string;
  completed_at?: string;
}

export type IssueCommentType =
  | "note"
  | "confirm"
  | "reject"
  | "assign"
  | "fix"
  | "verify"
  | "reopen"
  | "accept_risk";

export interface IssueComment {
  id: string;
  finding_id: string;
  comment_type: IssueCommentType;
  content: string;
  from_status?: string;
  to_status?: string;
  author_id: string;
  author_name?: string;
  created_at: string;
}

export interface FindingStatistics {
  total: number;
  by_severity: Record<string, number>;
  by_status: Record<string, number>;
  confirmed: number;
  false_positive: number;
  closed: number;
  confirmation_rate: number;
  false_positive_rate: number;
  closure_rate: number;
}

// ─── OpenCode Projects ────────────────────────────────────────────────────────

export interface CreateOpenCodeProjectPayload {
  name: string;
  description?: string;
  project_id?: string;
  code_path: string;
  selected_skills?: string[];
  selected_mcp_tools?: string[];
  audit_config?: Record<string, unknown>;
}

export interface UpdateOpenCodeProjectPayload {
  name?: string;
  description?: string;
  code_path?: string;
  selected_skills?: string[];
  selected_mcp_tools?: string[];
  audit_config?: Record<string, unknown>;
}

export interface AuditTriggerPayload {
  name?: string;
  description?: string;
  target_vulnerabilities?: string[];
  exclude_patterns?: string[];
  target_files?: string[];
  max_iterations?: number;
  timeout_seconds?: number;
}

export async function listOpenCodeProjects(params?: {
  skip?: number;
  limit?: number;
  status?: OpenCodeServerStatus;
}) {
  return apiClient.get<{ items: OpenCodeProject[]; total: number }>(
    "/opencode/projects",
    { params }
  );
}

export async function createOpenCodeProject(payload: CreateOpenCodeProjectPayload) {
  return apiClient.post<OpenCodeProject>("/opencode/projects", payload);
}

export async function getOpenCodeProject(projectId: string) {
  return apiClient.get<OpenCodeProject>(`/opencode/projects/${projectId}`);
}

export async function updateOpenCodeProject(
  projectId: string,
  payload: UpdateOpenCodeProjectPayload
) {
  return apiClient.patch<OpenCodeProject>(`/opencode/projects/${projectId}`, payload);
}

export async function deleteOpenCodeProject(projectId: string) {
  return apiClient.delete(`/opencode/projects/${projectId}`);
}

export async function startOpenCodeServer(projectId: string) {
  return apiClient.post<{
    message: string;
    port: number;
    pid?: number;
    status: string;
    simulated?: boolean;
  }>(`/opencode/projects/${projectId}/start`);
}

export async function stopOpenCodeServer(projectId: string) {
  return apiClient.post<{ message: string; status: string }>(
    `/opencode/projects/${projectId}/stop`
  );
}

export async function getServerStatus(projectId: string) {
  return apiClient.get<{
    status: OpenCodeServerStatus;
    port?: number;
    pid?: number;
    healthy: boolean;
    last_health_check?: string;
  }>(`/opencode/projects/${projectId}/status`);
}

export async function triggerAudit(projectId: string, payload: AuditTriggerPayload) {
  return apiClient.post<{
    message: string;
    agent_task_id: string;
    task_name: string;
    skills_count: number;
    mcp_tools_count: number;
  }>(`/opencode/projects/${projectId}/audit`, payload);
}

export async function getAuditHistory(projectId: string, params?: { skip?: number; limit?: number }) {
  return apiClient.get<{ items: OpenCodeAuditSession[]; total: number }>(
    `/opencode/projects/${projectId}/audit-history`,
    { params }
  );
}

// ─── Skills ────────────────────────────────────────────────────────────────────

export interface CreateSkillPayload {
  name: string;
  display_name: string;
  description?: string;
  category?: SkillCategory;
  system_prompt: string;
  tool_hints?: string[];
  vulnerability_types?: string[];
  severity_focus?: string[];
  icon?: string;
  tags?: string[];
}

export interface UpdateSkillPayload {
  display_name?: string;
  description?: string;
  category?: SkillCategory;
  system_prompt?: string;
  tool_hints?: string[];
  vulnerability_types?: string[];
  icon?: string;
  tags?: string[];
  is_active?: boolean;
}

export async function listSkills(params?: {
  category?: SkillCategory;
  is_active?: boolean;
  include_system?: boolean;
}) {
  return apiClient.get<{ items: OpenCodeSkill[]; total: number }>("/opencode/skills", { params });
}

export async function createSkill(payload: CreateSkillPayload) {
  return apiClient.post<OpenCodeSkill>("/opencode/skills", payload);
}

export async function getSkill(skillId: string) {
  return apiClient.get<OpenCodeSkill>(`/opencode/skills/${skillId}`);
}

export async function updateSkill(skillId: string, payload: UpdateSkillPayload) {
  return apiClient.patch<OpenCodeSkill>(`/opencode/skills/${skillId}`, payload);
}

export async function deleteSkill(skillId: string) {
  return apiClient.delete(`/opencode/skills/${skillId}`);
}

export async function previewSkillPrompt(skillId: string) {
  return apiClient.post<{ skill_id: string; system_prompt: string; combined_prompt: string }>(
    `/opencode/skills/${skillId}/preview`
  );
}

// ─── MCP Tools ────────────────────────────────────────────────────────────────

export interface CreateMCPToolPayload {
  name: string;
  display_name: string;
  description?: string;
  transport_type?: MCPTransportType;
  server_url?: string;
  command?: string;
  args?: string[];
  env_vars?: Record<string, string>;
  timeout_seconds?: number;
  capabilities?: string[];
  icon?: string;
}

export interface UpdateMCPToolPayload {
  display_name?: string;
  description?: string;
  transport_type?: MCPTransportType;
  server_url?: string;
  command?: string;
  args?: string[];
  env_vars?: Record<string, string>;
  timeout_seconds?: number;
  capabilities?: string[];
  is_active?: boolean;
}

export async function listMCPTools(params?: { is_active?: boolean }) {
  return apiClient.get<{ items: MCPToolConfig[]; total: number }>("/opencode/mcp-tools", { params });
}

export async function createMCPTool(payload: CreateMCPToolPayload) {
  return apiClient.post<MCPToolConfig>("/opencode/mcp-tools", payload);
}

export async function getMCPTool(toolId: string) {
  return apiClient.get<MCPToolConfig>(`/opencode/mcp-tools/${toolId}`);
}

export async function updateMCPTool(toolId: string, payload: UpdateMCPToolPayload) {
  return apiClient.patch<MCPToolConfig>(`/opencode/mcp-tools/${toolId}`, payload);
}

export async function deleteMCPTool(toolId: string) {
  return apiClient.delete(`/opencode/mcp-tools/${toolId}`);
}

export async function testMCPTool(toolId: string) {
  return apiClient.post<{ success: boolean; error?: string; tested_at: string }>(
    `/opencode/mcp-tools/${toolId}/test`
  );
}

// ─── Issue Management ─────────────────────────────────────────────────────────

export async function listFindingComments(findingId: string) {
  return apiClient.get<{ items: IssueComment[]; total: number }>(
    `/opencode/findings/${findingId}/comments`
  );
}

export interface AddCommentPayload {
  comment_type?: IssueCommentType;
  content: string;
  assignee_id?: string;
  due_date?: string;
}

export async function addFindingComment(findingId: string, payload: AddCommentPayload) {
  return apiClient.post<IssueComment>(`/opencode/findings/${findingId}/comments`, payload);
}

export interface ReviewFindingPayload {
  action: "confirm" | "reject" | "accept_risk" | "close" | "reopen";
  notes?: string;
  assignee_id?: string;
  due_date?: string;
}

export async function reviewFinding(findingId: string, payload: ReviewFindingPayload) {
  return apiClient.patch(`/opencode/findings/${findingId}/review`, payload);
}

export interface BatchUpdatePayload {
  finding_ids: string[];
  status?: string;
  assignee_id?: string;
  due_date?: string;
}

export async function batchUpdateFindings(payload: BatchUpdatePayload) {
  return apiClient.post<{ updated: number; total: number }>(
    "/opencode/findings/batch-update",
    payload
  );
}

export async function getFindingsStatistics(taskId: string) {
  return apiClient.get<FindingStatistics>(
    `/opencode/tasks/${taskId}/findings/statistics`
  );
}
